### PR TITLE
[QNN EP] Support Resize's pytorch_half_pixel coordinate transformation mode on HTP

### PR DESCRIPTION
**Pull request by adrianlizarraga**
### Description
- Now uses QNN's Resize operator for quantized models
- Still uses QNN's ResizeBilinear or ResizeNearestNeighbor for non-quantized models.

### Motivation and Context
This update is necessary to support more models on QNN HTP backend. Specifically, we need to support Resize's `pytorch_half_pixel` coordinate transformation mode on HTP.

### CODE REVIEW

File: onnxruntime/core/providers/qnn/builder/opbuilder/resize_op_builder.cc [hunk 1]
```diff
@@ -38,51 +42,227 @@ class ResizeOpBuilder : public BaseOpBuilder {
...
+    std::string msg = coordinate_mode + " mode not supported, QNN Resize only supports align_corners and half_pixel";
     return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, msg);
   }
 
-  if ("nearest" == resize_mode) {
-    const std::string& nearest_mode = node_helper.Get("nearest_mode", "round_prefer_floor");
-    ORT_RETURN_IF_NOT("floor" == nearest_mode, "QNN Resize only support nearest_mode: floor!");
+  if (resize_mode == "nearest") {
+    const std::string& nearest_mode = GetOnnxAttr(node_helper, onnx_nearest_mode_attr);
+    ORT_RETURN_IF_NOT("round_prefer_floor" == nearest_mode, "QNN Resize only supports nearest_mode: round_prefer_floor!");
```
Comment (adrianlizarraga) [inline onnxruntime/core/providers/qnn/builder/opbuilder/resize_op_builder.cc:216]: Unit tests revealed that QNN's ResizeNearestNeighbor op actually uses `round_prefer_floor`.

File: onnxruntime/core/providers/qnn/builder/opbuilder/resize_op_builder.cc [hunk 2]
```diff
@@ -38,51 +42,252 @@ class ResizeOpBuilder : public BaseOpBuilder {
...
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN Resize only supports 4D!");
+  }
+
+  const auto& output_0 = node_unit.Outputs()[0];
+  std::vector<uint32_t> output_shape;
+  ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(output_0.node_arg, output_shape),
+                    "QNN EP: Cannot get output shape for Resize op");
+  if (output_shape.size() != 4) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN Resize only supports 4D!");
   }
```
Comment (HectorSVC) [inline onnxruntime/core/providers/qnn/builder/opbuilder/resize_op_builder.cc:249]: use ORT_RETURN_IF

File: onnxruntime/core/providers/qnn/builder/opbuilder/resize_op_builder.cc [hunk 3]
```diff
@@ -38,51 +42,252 @@ class ResizeOpBuilder : public BaseOpBuilder {
...
+    const std::string& nearest_mode = GetOnnxAttr(node_helper, onnx_nearest_mode_attr);
+    ORT_RETURN_IF_NOT("floor" == nearest_mode, "QNN Resize only supports nearest_mode: floor!");  // This is wrong!
   }
 
   auto& input_0 = node_unit.Inputs()[0];
   std::vector<uint32_t> input_shape;
-  ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(input_0.node_arg, input_shape), "Cannot get shape");
+  ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(input_0.node_arg, input_shape),
+                    "QNN EP: Cannot get input shape for Resize op");
   if (input_shape.size() != 4) {
```
Comment (HectorSVC) [inline onnxruntime/core/providers/qnn/builder/opbuilder/resize_op_builder.cc:239]: >if (input_shape.size() != 4) { [](start = 2, length = 30)

use ORT_RETURN_IF

File: onnxruntime/core/providers/qnn/builder/opbuilder/resize_op_builder.cc (hunk 2 above)
Comment (HectorSVC) [reply onnxruntime/core/providers/qnn/builder/opbuilder/resize_op_builder.cc:249]: merge it with if (input_shape.size() != 4)
ORT_RETURN_IF(input_shape.size() != 4 || output_shape.size() != 4, "QNN Resize only supports 4D!");

---
In reply to: [1160306095](https://github.com/microsoft/onnxruntime/pull/15390#discussion_r1160306095) [](ancestors = 1160306095)

File: onnxruntime/core/providers/qnn/builder/opbuilder/resize_op_builder.cc [hunk 4]
```diff
@@ -38,51 +42,252 @@ class ResizeOpBuilder : public BaseOpBuilder {
...
+  std::vector<uint32_t> output_shape;
+  ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(output_0.node_arg, output_shape),
+                    "QNN EP: Cannot get output shape for Resize op");
+  if (output_shape.size() != 4) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN Resize only supports 4D!");
   }
 
   ONNX_NAMESPACE::DataType input_data_type = input_0.node_arg.Type();
-  if (!is_quantized_model && input_data_type != ONNX_NAMESPACE::Utils::DataTypeUtils::ToType("float")) {
+  if (input_data_type != ONNX_NAMESPACE::Utils::DataTypeUtils::ToType("float")) {
```
Comment (HectorSVC) [inline onnxruntime/core/providers/qnn/builder/opbuilder/resize_op_builder.cc:252]: >if (input_data_type != ONNX_NAMESPACE::Utils::DataTypeUtils::ToType("float")) [](start = 2, length = 77)

use ORT_RETURN_IF

**Review by HectorSVC — APPROVED** [review]
:shipit:

**Pull request merged**